src/mscorlib/src/System/Lazy.cs

```diff
@@ -202,27 +308,123 @@ public Lazy(Func<T> valueFactory, bool isThreadSafe)
-                return LazyHelpers.PUBLICATION_ONLY_SENTINEL;
-            else if (mode != LazyThreadSafetyMode.None)
-                throw new ArgumentOutOfRangeException(nameof(mode), Environment.GetResourceString("Lazy_ctor_ModeInvalid"));
-            
-            return null; // None mode
+            _value = CreateViaDefaultConstructor();
+            _state = null;
+
+            return _value;
+        }
+
+        private T ViaFactory(LazyThreadSafetyMode mode)
+        {
+            try
+            {
+                var factory = _factory;
+                if (factory == null)
+                    throw new InvalidOperationException(Environment.GetResourceString("Lazy_Value_RecursiveCallsToValue"));
+                _factory = null;
+
+                _value = factory();
+                _state = null;
+
+                return _value;
+            }
+            catch (Exception exception)
+            {
+                _state = new LazyNonGeneric(mode, exception);
+                throw;
+            }
+        }
+
+        private T ExecutionAndPublication(LazyNonGeneric executionAndPublication, bool useDefaultConstructor)
+        {
+            lock (executionAndPublication)
+            {
+                // it's possible for multiple calls to have piled up behind the lock, so we need to check
+                // to see if the ExecutionAndPublication object is still the current implementation.
+                if (ReferenceEquals(_state, executionAndPublication))
+                    return useDefaultConstructor ? ViaConstructor() : ViaFactory(LazyThreadSafetyMode.ExecutionAndPublication);
+
+                return Value;
+            }
+        }
+
+        private T PublicationOnly(LazyNonGeneric publicationOnly, T possibleValue)
+        {
+            try { }
+            finally
+            {
+                // we run this in a finally block to ensure that we don't get a partial completion due
+                // to a Thread.Abort, which could mean that other threads might be left in infinite loops
+                var previous = Interlocked.CompareExchange(ref _state, LazyNonGeneric.PublicationOnlySpinWait, publicationOnly);
+                if (previous == publicationOnly)
+                {
+                    _value = possibleValue;
+                    _factory = null;
+                    _state = null;
+                }
+            }
+
+            return Value;
+        }
+
+        private T PublicationOnlyViaConstructor(LazyNonGeneric initializer)
+        {
+            return PublicationOnly(initializer, CreateViaDefaultConstructor());
+        }
+
+        private T PublicationOnlyViaFactory(LazyNonGeneric initializer)
+        {
+            var factory = _factory;
+            if (factory == null)
+                return PublicationOnlySpinWait();
+            return PublicationOnly(initializer, factory());
+        }
+
+        private T PublicationOnlySpinWait()
```

Comment:
Nit: maybe rename this to something like `PublicationOnlyWaitForOtherThreadToPublish`?